File: spine/orreth_spine/digest.py
# PROVENANCE: Claude Fable 5.1 (claude-fable-5-1) — rearch P5 sp3, the Digest (MEM-3) · 2026-09-19
"""The Digest (canon 0003, JB's compression law): the short version of an
episode, written at its boundary, CITING every record it compresses. v0
is extractive and deterministic — the head of every exchange, who
replied and when, the words acquired — so a rebuilt digest is
byte-identical when nothing changed, and a grown episode gets a SIBLING
digest that supersedes (never an overwrite). A mind's prose digest can
land later as a sibling of the same record. The pack reads the short
version first; the verbatim opens on demand through the recall door."""
from __future__ import annotations

import json
import secrets

from . import envelope as ev, outbox

DIGEST_EVENT = "orreth.digest.landed.v1"
HEAD = 120


def ensure_schema(conn) -> None:
    from .outbox import once
    if not once(conn, "digest"):
        return
    with conn.transaction():
        cur = conn.cursor()
        cur.execute("SELECT pg_advisory_xact_lock(742199)")  # DDL race guard
        cur.execute(
            "CREATE TABLE IF NOT EXISTS spine_digests ("
            " digest_id text PRIMARY KEY, kind text NOT NULL, ref text NOT NULL,"
            " body text NOT NULL, sources text NOT NULL, hash text NOT NULL,"
            " by_did text NOT NULL, scope text NOT NULL, supersedes text,"
            " built_at timestamptz NOT NULL DEFAULT now(),"
            " valid_from timestamptz NOT NULL DEFAULT now(), valid_to timestamptz)")
        cur.execute("CREATE UNIQUE INDEX IF NOT EXISTS spine_digests_current"
                    " ON spine_digests (kind, ref, scope) WHERE valid_to IS NULL")


def _head(text: str, n: int = HEAD) -> str:
    text = " ".join((text or "").split())
    return text if len(text) <= n else text[:n - 1].rstrip() + "…"


def compose_session(conn, session_id: str) -> tuple[str, list[str]] | None:
    """The short version of a session, from the Record alone: every ask's
    head, who replied and when, the reply's head; the words acquired in
    the session's span. Returns (body, sources) or None if no session."""
    from .resident import ensure_schema as _ground
    _ground(conn)
    cur = conn.cursor()
    cur.execute("SELECT session_id, title, opened_at, person FROM spine_sessions"
                " WHERE session_id = %s AND scope = %s", (session_id, ev.scope()))
    row = cur.fetchone()
    if row is None:
        return None
    _sid, title, opened, person = row
    # the episode's span: from this session's opening to the next session's
    # opening (the same human, this world) — or now, for the latest one
    cur.execute("SELECT coalesce(min(opened_at), now()) FROM spine_sessions"
                " WHERE person = %s AND scope = %s AND opened_at > %s",
                (person, ev.scope(), opened))
    span_end = cur.fetchone()[0]
    cur.execute(
        "SELECT a.ask_id, a.text, a.reply, a.status, a.asked_at, a.replied_at,"
        " coalesce(j.name, 'a resident')"
        " FROM spine_asks a LEFT JOIN LATERAL ("
        "  SELECT name FROM spine_joins WHERE did = a.served_by"
        "  ORDER BY join_id DESC LIMIT 1) j ON true"
        " WHERE a.session = %s ORDER BY a.asked_at", (session_id,))
    asks = cur.fetchall()
    sources = [a[0] for a in asks]
    lines = [f"session {session_id[4:10]}" + (f" — {title}" if title else "")
             + f" · opened {opened.strftime('%a %b %d %H:%M')} · {len(asks)} ask"
             + ("" if len(asks) == 1 else "s")]
    for aid, text, reply, status, at, rat, who in asks:
        lines.append(f"· {at.strftime('%H:%M')} asked: {_head(text)}")
        if status == "replied" and reply:
            lines.append(f"  {who} replied ({rat.strftime('%H:%M') if rat else '—'}): {_head(reply)}")
        elif status != "received":
            lines.append(f"  {status}")
    if asks:
        cur.execute("SELECT to_regclass('spine_memories') IS NOT NULL")
        if cur.fetchone()[0]:
            cur.execute(
                "SELECT namespace, key, body FROM spine_memories WHERE scope = %s"
                " AND landed_at BETWEEN %s AND %s ORDER BY landed_at",
                (ev.scope(), opened, span_end))
            for ns, key, body in cur.fetchall():
                sources.append(f"{ns}/{key}")
                lines.append(f"· acquired [{ns}/{key}]: {_head(body)}")
    return "\n".join(lines), sources


def build(conn, session_id: str, by: str = "the digest builder") -> dict | None:
    """Write (or rebuild) the session's digest: the same substance lands
    nothing new; changed substance lands a sibling that supersedes."""
    ensure_schema(conn)
    outbox.ensure_schema(conn)
    made = compose_session(conn, session_id)
    if made is None:
        return None
    body, sources = made
    h = ev.content_hash(body)
    with conn.transaction():
        cur = conn.cursor()
        cur.execute("SELECT digest_id, hash FROM spine_digests WHERE kind = 'session'"
                    " AND ref = %s AND scope = %s AND valid_to IS NULL FOR UPDATE",
                    (session_id, ev.scope()))
        prev = cur.fetchone()
        if prev and prev[1] == h:
            return {"digest_id": prev[0], "hash": h, "body": body, "sources": sources,
                    "new": False}
        did = "dig_" + secrets.token_hex(5)
        if prev:
            cur.execute("UPDATE spine_digests SET valid_to = now() WHERE digest_id = %s",
                        (prev[0],))
        cur.execute(
            "INSERT INTO spine_digests (digest_id, kind, ref, body, sources, hash,"
            " by_did, scope, supersedes) VALUES (%s, 'session', %s, %s, %s, %s, %s, %s, %s)",
            (did, session_id, body, json.dumps(sources), h, by, ev.scope(),
             prev[1] if prev else None))
        e = ev.make_envelope(
            kind="event", type=DIGEST_EVENT, universe_id=ev.scope(), scope_path=ev.scope(),
            payload={"ref": did, "hash": h, "session": session_id,
                     "sources": len(sources)}, correlation_id=session_id,
            authority_chain=[by])
        outbox.add_row(cur, ev.encode(e), e["message_id"])
    return {"digest_id": did, "hash": h, "body": body, "sources": sources, "new": True}


def of_session(conn, session_id: str) -> dict | None:
    """The current digest of a session, with its sources."""
    ensure_schema(conn)
    cur = conn.cursor()
    cur.execute("SELECT digest_id, body, sources, hash, by_did, built_at, supersedes"
                " FROM spine_digests WHERE kind = 'session' AND ref = %s AND scope = %s"
                " AND valid_to IS NULL", (session_id, ev.scope()))
    r = cur.fetchone()
    return {"digest_id": r[0], "session": session_id, "body": r[1],
            "sources": json.loads(r[2]), "hash": r[3], "by": r[4],
            "built_at": r[5].isoformat(), "supersedes": r[6]} if r else None


def for_person(conn, person: str, *, exclude: str | None = None,
               window: dict | None = None, limit: int = 3) -> list[dict]:
    """The short version first (the pack's third rung): the current digests
    of this human's sessions in this world — newest first; inside a
    window, only sessions with asks in it."""
    ensure_schema(conn)
    cur = conn.cursor()
    if window and window.get("from") and window.get("to"):
        cur.execute(
            "SELECT d.digest_id, d.ref, d.body, d.sources FROM spine_digests d"
            " JOIN spine_sessions s ON s.session_id = d.ref"
            " WHERE d.kind = 'session' AND d.scope = %s AND d.valid_to IS NULL"
            " AND s.person = %s AND d.ref <> %s AND EXISTS (SELECT 1 FROM spine_asks a"
            "  WHERE a.session = d.ref AND a.asked_at BETWEEN %s AND %s)"
            " ORDER BY s.opened_at DESC LIMIT %s",
            (ev.scope(), person, exclude or "", window["from"], window["to"], limit))
    else:
        cur.execute(
            "SELECT d.digest_id, d.ref, d.body, d.sources FROM spine_digests d"
            " JOIN spine_sessions s ON s.session_id = d.ref"
            " WHERE d.kind = 'session' AND d.scope = %s AND d.valid_to IS NULL"
            " AND s.person = %s AND d.ref <> %s ORDER BY s.opened_at DESC LIMIT %s",
            (ev.scope(), person, exclude or "", limit))
    return [{"digest_id": r[0], "session": r[1], "body": r[2],
             "sources": json.loads(r[3])} for r in cur.fetchall()]

File: backend/plane/crates/orreth-spine/src/digest.rs
// PROVENANCE: Claude Fable 5.1 (claude-fable-5-1) — rearch P7 sp5, memory and the export · 2026-09-24
// Amended: Claude Fable 5.1 (claude-fable-5-1) — rearch walk #13 cures: W51 a roll is a fact the feed carries · W52 the digest in the human's zone · THE GUIDE door · 2026-09-24
//! The Digest — `orreth_spine.digest` ported (MEM-3, canon 0003): the short
//! version of a session composed from the Record alone (deterministic,
//! extractive — a rebuild lands byte-identical), the sibling law over
//! `spine_digests` (the same substance lands nothing; changed substance
//! supersedes), its fact through the outbox, the current digest of a
//! session, the rebuild after a purge, and the pack's third rung — the
//! same SQL as the reference so both kernels write one Digest.

use crate::envelope::{self, Mint};
use crate::ground::Ground;
use crate::hash::content_hash;
use crate::memory::{digest_lines, digest_payload, Civil, DigestAsk, DIGEST_EVENT};
use crate::outbox;
use crate::schema::has_table;
use crate::world::{isoformat, token_hex, RoadError};
use serde_json::{json, Value};
use std::time::SystemTime;

pub const BUILDER: &str = "the digest builder";

/// `digest.compose_session`: (body, sources, state) — None when no such session.
pub async fn compose_session(
    g: &Ground,
    scope: &str,
    session_id: &str,
    zone: &str,
) -> Result<Option<(String, Vec<String>, String)>, RoadError> {
    // W52: the digest's clock is the human's — the ground shifts every time into the zone
    // (`AT TIME ZONE`, the same tzdata the Python spine's ZoneInfo reads) before it is worded
    let row = g
        .client()
        .query_opt(
            "SELECT session_id, title, opened_at, person, coalesce(state, 'in'), \
             to_char(opened_at AT TIME ZONE $3, 'YYYY-MM-DD\"T\"HH24:MI:SS') FROM spine_sessions \
             WHERE session_id = $1 AND scope = $2",
            &[&session_id, &scope, &zone],
        )
        .await?;
    let Some(row) = row else {
        return Ok(None);
    };
    let title: Option<String> = row.get(1);
    let opened: SystemTime = row.get(2);
    let person: String = row.get(3);
    let state: String = row.get(4);
    let opened_civil = Civil::parse(&row.get::<_, String>(5)).expect("the ground's own text");
    let span_end: SystemTime = g
        .client()
        .query_one(
            "SELECT coalesce(min(opened_at), now()) FROM spine_sessions WHERE person = $1 AND scope = $2 \
             AND opened_at > $3",
            &[&person, &scope, &opened],
        )
        .await?
        .get(0);
    let asks = g
        .client()
        .query(
            "SELECT a.ask_id, a.text, a.reply, a.status, \
             to_char(a.asked_at AT TIME ZONE $2, 'YYYY-MM-DD\"T\"HH24:MI:SS'), \
             to_char(a.replied_at AT TIME ZONE $2, 'YYYY-MM-DD\"T\"HH24:MI:SS'), coalesce(j.name, 'a \
             resident') FROM spine_asks a LEFT JOIN LATERAL (  SELECT name FROM spine_joins WHERE did = \
             a.served_by  ORDER BY join_id DESC LIMIT 1) j ON true WHERE a.session = $1 ORDER BY a.asked_at",
            &[&session_id, &zone],
        )
        .await?;
    let asks: Vec<DigestAsk> = asks
        .iter()
        .map(|r| DigestAsk {
            ask_id: r.get(0),
            text: r.get(1),
            reply: r.get(2),
            status: r.get(3),
            asked_at: Civil::parse(&r.get::<_, String>(4)).expect("the ground's own text"),
            replied_at: r.get::<_, Option<String>>(5).and_then(|t| Civil::parse(&t)),
            who: r.get(6),
        })
        .collect();
    let mut memories: Vec<(String, String, String)> = Vec::new();
    if !asks.is_empty() && has_table(g, "spine_memories").await? {
        let rows = g
            .client()
            .query(
                "SELECT namespace, key, body FROM spine_memories WHERE scope = $1 AND state = $2 AND \
                 landed_at BETWEEN $3 AND $4 ORDER BY landed_at",
                &[&scope, &state, &opened, &span_end],
            )
            .await?;
        memories = rows
            .iter()
            .map(|r| (r.get(0), r.get(1), r.get(2)))
            .collect();
    }
    let (body, sources) =
        digest_lines(session_id, title.as_deref(), opened_civil, &asks, &memories);
    Ok(Some((body, sources, state)))
}

/// `digest.build`: write (or rebuild) the session's digest — the same
/// substance lands nothing new; changed substance lands a sibling.
pub async fn build(
    g: &mut Ground,
    scope: &str,
    session_id: &str,
    by: &str,
    zone: &str,
) -> Result<Option<Value>, RoadError> {
    let Some((body, sources, state)) = compose_session(g, scope, session_id, zone).await? else {
        return Ok(None);
    };
    let h = content_hash(&Value::String(body.clone()));
    let tx = g.client_mut().transaction().await?;
    let prev = tx
        .query_opt(
            "SELECT digest_id, hash FROM spine_digests WHERE kind = 'session' AND ref = $1 AND scope = $2 \
             AND valid_to IS NULL FOR UPDATE",
            &[&session_id, &scope],
        )
        .await?;
    let prev: Option<(String, String)> = prev.map(|r| (r.get(0), r.get(1)));
    if let Some((id, ph)) = &prev {
        if *ph == h {
            tx.commit().await?;
            return Ok(Some(
                json!({"digest_id": id, "hash": h, "body": body, "sources": sources, "new": false}),
            ));
        }
    }
    let did = format!("dig_{}", token_hex(5));
    if let Some((id, _)) = &prev {
        tx.execute(
            "UPDATE spine_digests SET valid_to = now() WHERE digest_id = $1",
            &[id],
        )
        .await?;
    }
    let sources_json = serde_json::to_string(&sources).unwrap_or_else(|_| "[]".into());
    tx.execute(
        "INSERT INTO spine_digests (digest_id, kind, ref, body, sources, hash, by_did, scope, supersedes, \
         state) VALUES ($1, 'session', $2, $3, $4, $5, $6, $7, $8, $9)",
        &[&did, &session_id, &body, &sources_json, &h, &by, &scope, &prev.as_ref().map(|p| p.1.clone()), &state],
    )
    .await?;
    let e = Mint {
        kind: "event".into(),
        r#type: DIGEST_EVENT.into(),
        universe_id: scope.into(),
        scope_path: scope.into(),
        payload: digest_payload(&did, &body, session_id, sources.len()),
        correlation_id: Some(session_id.into()),
        authority_chain: Some(vec![by.to_string()]),
        aggregate: None,
        marker: None,
    }
    .mint()?;
    let raw = envelope::encode(&e)?;
    outbox::add_row(&tx, &raw, e["message_id"].as_str().unwrap_or_default()).await?;
    tx.commit().await?;
    Ok(Some(
        json!({"digest_id": did, "hash": h, "body": body, "sources": sources, "new": true}),
    ))
}

/// The current digest of a session, with its sources.
pub async fn of_session(
    g: &Ground,
    scope: &str,
    session_id: &str,
) -> Result<Option<Value>, RoadError> {
    let row = g
        .client()
        .query_opt(
            "SELECT digest_id, body, sources, hash, by_did, built_at, supersedes FROM spine_digests WHERE \
             kind = 'session' AND ref = $1 AND scope = $2 AND valid_to IS NULL",
            &[&session_id, &scope],
        )
        .await?;
    Ok(row.map(|r| {
        json!({
            "digest_id": r.get::<_, String>(0), "session": session_id, "body": r.get::<_, String>(1),
            "sources": serde_json::from_str::<Value>(&r.get::<_, String>(2)).unwrap_or(json!([])),
            "hash": r.get::<_, String>(3), "by": r.get::<_, String>(4),
            "built_at": isoformat(r.get::<_, SystemTime>(5)), "supersedes": r.get::<_, Option<String>>(6),
        })
    }))
}

/// After a purge: every current digest that cited the ref is rebuilt.
pub async fn rebuild_citing(
    g: &mut Ground,
    scope: &str,
    r#ref: &str,
    by: &str,
    zone: &str,
) -> Result<usize, RoadError> {
    let needle = format!("%\"{}\"%", r#ref);
    let rows = g
        .client()
        .query(
            "SELECT ref FROM spine_digests WHERE kind = 'session' AND scope = $1 AND valid_to IS NULL AND \
             sources LIKE $2",
            &[&scope, &needle],
        )
        .await?;
    let sessions: Vec<String> = rows.iter().map(|r| r.get(0)).collect();
    let mut n = 0;
    for sid in sessions {
        if let Some(made) = build(g, scope, &sid, by, zone).await? {
            if made["new"].as_bool().unwrap_or(false) {
                n += 1;
            }
        }
    }
    Ok(n)
}

/// A listed session without its short version gets it from the log (MEM-3 —
/// the Digest is a projection, rebuildable at any time).
pub async fn build_missing(
    g: &mut Ground,
    scope: &str,
    person: &str,
    limit: i64,
    zone: &str,
) -> Result<usize, RoadError> {
    let rows = g
        .client()
        .query(
            "SELECT s.session_id FROM spine_sessions s WHERE s.person = $1 AND s.scope = $2 AND EXISTS (SELECT \
             1 FROM spine_asks a WHERE a.session = s.session_id) AND NOT EXISTS (SELECT 1 FROM spine_digests d \
             WHERE d.kind = 'session'  AND d.ref = s.session_id AND d.scope = s.scope AND d.valid_to IS NULL) \
             ORDER BY s.opened_at DESC LIMIT $3",
            &[&person, &scope, &limit],
        )
        .await?;
    let sids: Vec<String> = rows.iter().map(|r| r.get(0)).collect();
    let mut n = 0;
    for sid in sids {
        build(g, scope, &sid, BUILDER, zone).await?;
        n += 1;
    }
    Ok(n)
}
